MHD

guadabsb15 opened the pull request:
So far it looks like the MHD solver works in the absence of B fields

zingale committed:
fix indexing of aux

zingale committed:
indent

zingale committed:
use temp as initial guess for EOS

zingale committed:
fix UFX

zingale committed:
initial q(QTEMP)

zingale committed:
pass T into MHD soundspeed to have good initial guess

guadabsb15 committed:
Initialize Ip and Im, now valgrind doesn't complain about using unitialized things

zingale committed:
add some comments

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

guadabsb15 committed:
fixed mixed indices when calling conervative variables with primitive indices

zingale committed:
index + define species (hackishly) on interfaces

zingale committed:
fix the setting of interface species -- these are primitive variables

zingale committed:
fix bounds issue

zingale committed:
fill species

guadabsb15 committed:
fix the Jacobian in qflux

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

zingale committed:
a derivation of the Jacobian b/w conserved to primitive

guadabsb15 committed:
added timestep estimator for mhd
some changes in eos_soundspeed_mhd

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

guadabsb15 committed:
changed my mind... modified interface of enforce_consistent_e to include there
the magnetic e term if using MHD

guadabsb15 committed:
removed the function I had made to add magnetic_e
moved the initializing magnetic fields part, so when enforce_consistent_e
is called, that information is passed

guadabsb15 committed:
handle magnetic e term in reset_internal_e

zingale committed:
add comment

zingale committed:
fix Jacobian

zingale committed:
uncomment advection_util

guadabsb15 committed:
fix #ifdef MH for MHD and removed the correction in two places were it wasn't needed

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

zingale committed:
fix indexing

guadabsb15 committed:
fixed to cell center Magnetic fields

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

zingale committed:
use a left and right gamma_1

zingale committed:
don't include the magnetic pressure in thermodynamics when doing eos_input_rp

guadabsb15 committed:
first pass removing magentic term of pressure when used for primitive variables
and adding the magnetic term when getting fluxes

zingale committed:
Merge branch 'development' into MHD

zingale committed:
add some todos + 1 fixed of left/right states

guadabsb15 committed:
comment out computation of cell center magnetic fields in reset_internal_e in Castro_util
since there was a problem when running after compiling in DEBUG mode (mismatch of indices)
For now, it just assigns 0.5 for those variables... sketchy, but the run finishes. Will fix this soon

guadabsb15 committed:
modified makfile to have by default DEBUG=TRUE and USE_MPI=FALSE

guadabsb15 committed:
deleted inputs and probin that aren't used

guadabsb15 committed:
added setup problem with B=0

guadabsb15 committed:
not relying on the assumption of variables and order between "URHO:UEDEN", so instead of that... assigning for each one individually

guadabsb15 committed:
Do a call to eos to determine T before the calls to soundspeed_mhd so the calls inside that subroutine can be with eos_input_rt

guadabsb15 committed:
fixed an index that was QRHO instead of QPRES

guadabsb15 committed:
removed TODO that were already done

guadabsb15 committed:
fixed sign error

guadabsb15 committed:
forgot to fix sign error for right state

guadabsb15 committed:
checked HLLD paper to determine when it is used total p and when it is gas p

guadabsb15 committed:
set the interface states for rho eint + some comments in ct_upwind

guadabsb15 committed:
fix in mhd_sound_speed

guadabsb15 committed:
removed advection_util_nd and replaced it by mhd_util, where currently just had enforce_minimum_density

fix in the timestep estimation

guadabsb15 committed:
fix some flux array sizes from QVAR to NVAR+3 and in electric.f90, fixed mixed indices of conservative with primitive

guadabsb15 committed:
another pass to the riemann solver, comparing with paper

guadabsb15 committed:
some clean up to not assume URHO and UEDEN are contiguous

guadabsb15 committed:
added inputs for the sod in different directions

guadabsb15 committed:
added idir = 2 and 3 in the initialization

zingale committed:
fix indexing

guadabsb15 committed:
fix in left and right eigenvectors for y and z

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

guadabsb15 committed:
Some cleanup in Castro_advection_mhd and change to not tile

guadabsb15 committed:
adding sedov problem to test with mhd solver

guadabsb15 committed:
fix in parenthesis in plm and add gamma to probin in Sedov

zingale committed:
add a note here about defining the grown B stuff

guadabsb15 committed:
Defined grown B*_old_tmp in Castro_advance for call to expand_state

guadabsb15 committed:
Initializing magnetic field for idir 1 and 2

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
change in Make.package to not include trace_ppm at the moment. Fluxes correction and fix index in one call of magnetic field

zingale committed:
this is our new reference case

guadabsb15 committed:
some changes in the bs of the eigensystem, include B component in enforce_minimum_density and added Ecorrection subroutine
Syncing with Nyx

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

guadabsb15 committed:
inputs for BrioWu in x

guadabsb15 committed:
fix indices that were mixing components

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
no MOL for now

guadabsb15 committed:
Merge branch 'development' into MHD

zingale commented on Source/driver/_cpp_parameters:
```diff
@@ -76,6 +76,10 @@ hybrid_hydro                 int           0                  y
 ppm_type                     int           1                  y
 
+#Lupe adding for MHD thing
+ppm_reference                int           1                  y             MHD
```
we'll eventually want to remove these -- they are old features

zingale commented on Source/mhd/Castro_advection_mhd_3d.F90:
```diff
@@ -0,0 +1,905 @@
+
+
+           eos_state % rho = q(i,j,k,QRHO)
+           eos_state % e   = q(i,j,k, QREINT)
```
are we sure that q(i,j,k,QREINT) is not (rho e) here?

zingale replied:
okay... it looks like it is just e (as shown below)

zingale commented on Source/mhd/ct_upwind.f90:
```diff
@@ -0,0 +1,1312 @@
+										((Ex(i,j+1,k+1) - Ex(i,j,k+1)) + &
+										 (Ex(i,j+1,k  ) - Ex(i,j,k)))
+				!-> Affected by Z flux
+				uL(i,j,k,UMAGX,1,2) = um(i,j,k,UMAGX,1) - dt/(3.d0*dx)*(Ey(i,j,k+1) - Ey(i,j,k))
```
shouldn't this be `uL(i,j,k,UMAGX,1,1)`, instead of `...1,2)`?

zingale replied:
actually, I don't understand what that last index is for

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

guadabsb15 committed:
Working on species. Still need to work on qflux  in ct_upwind.f90 and in the riemann solver

guadabsb15 committed:
more species work. Test is coming soon

guadabsb15 committed:
starting setup to test species

zingale committed:
switch net, some init

zingale committed:
define T

guadabsb15 committed:
fix limits

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

guadabsb15 committed:
addressing out of bound issue found while running with DEBUG

In plm was accessing lo-1, lo+1 ...etc, did a temp_s similar to the ones there for the other variables
of size (s_lo-1, s_hi+1)

guadabsb15 committed:
fixed dx factor on plm for species and periodic b.c. on the inputs for the species test

guadabsb15 committed:
started adding slope options for plm

guadabsb15 committed:
changed electric fields, according to Stone paper

guadabsb15 committed:
Add option of slope 0 in plm. Fix in Castro_F.H repetition in ca_face_fillx header

guadabsb15 committed:
don't use mhd_enforce_minimum_density for now

guadabsb15 committed:
fix in internal energy

guadabsb15 committed:
another fix in energy

guadabsb15 committed:
the previous one was not a fix. Now back to rho*eint

guadabsb15 committed:
Change in normal component of field in plm to use face centered
Removed a call to EOS in Castro_advection_mhd

guadabsb15 committed:
fix in reset_internal_e

guadabsb15 committed:
filled UTEMP in Prob_3d.f90 and added small_temp and small_dens to inputs

guadabsb15 committed:
change in BC

guadabsb15 committed:
fix index

guadabsb15 committed:
another index fix

guadabsb15 committed:
testing Brio & Wu for different directions

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
add wd merger setup and included face_fill* subroutines to work with dimension agnostic
The setup compiles

guadabsb15 committed:
Removed grav associated just with mhd and the respective add_grav subroutine. Also removed duplicated FillPatchAdd

guadabsb15 committed:
forgot to add changes on the makefile

guadabsb15 committed:
added Cons to prim for sources, and sending it to plm

guadabsb15 committed:
addressing fluxes and mass_fluxes

guadabsb15 committed:
adding src to plm

guadabsb15 committed:
clean up - just one do loop getting slope rather than two

zingale committed:
fix initialization of B

zingale committed:
fix dimensioning:

guadabsb15 committed:
added headers of face_fill_* to work with agnostic version

guadabsb15 committed:
add initial guess in Temperature before call eos with eos_input_rp

guadabsb15 committed:
initialize Bout

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
replaced bl_error and bl_constants by amrex_error and amrex_constants

guadabsb15 committed:
changes from bl_ to amrex_

guadabsb15 committed:
more fix B initialization

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
compiles without MHD

guadabsb15 committed:
compiles when using MHD too

guadabsb15 committed:
working on getting flattening

guadabsb15 committed:
more on flattening

guadabsb15 committed:
change in loop limits in plm and attempt on using Ptotal for flattening

guadabsb15 committed:
added second order MC limiter

guadabsb15 committed:
change dx by dy and dz for y and z direction

guadabsb15 committed:
adding Alfven wave test for wavenumber n=(1,0) following Crockett et.al 2004

guadabsb15 committed:
fix in index

guadabsb15 committed:
Merge branch 'MHD' of https://github.com/AMReX-Astro/Castro into MHD

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
compiles after merge

guadabsb15 committed:
add in the inputs option to not use flattening

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
fix in hlld and some likely fixes in ct_upwind

guadabsb15 committed:
moved back a few indices

guadabsb15 committed:
change y direction indices in corner couple to be consistent with
hydro corner couple

guadabsb15 committed:
update magentic fields consistent with paper

guadabsb15 committed:
Merge branch 'development' into MHD

Commit:
Merge branch 'development' into MHD

guadabsb15 committed:
compiles with USE_MHD=FALSE

guadabsb15 committed:
now it compiles for USE_MHD=TRUE still need to check if it runs

guadabsb15 committed:
Sod problem runs and results agree with previous one

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
more progress for compiling

guadabsb15 committed:
Sod problem on Exec/mhd_tests now compiles and runs.
Need to update initialization for the rest of setups in mhd_tests

guadabsb15 committed:
Sedov now compiles and runs too
Note that during the merge seems like MPI stopped working when using the
MHD solver

guadabsb15 committed:
Sedov and species compile and run now. BrioWu compiles but is not
running

guadabsb15 committed:
fix in facefilly and facefillz

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
fix compilation

guadabsb15 committed:
fixed signs to be consistent with GS05 and GS07

guadabsb15 committed:
alfven problem and other inputs updates

guadabsb15 committed:
Add Dai-Woodward test and modified some previous inputs

guadabsb15 committed:
added Orszag-Tang test

guadabsb15 committed:
A pass through the eigensystem.
Brio-Wu in y and z directions looks consistent now with x-direction

guadabsb15 committed:
Add fast rarefaction test as in Miniati (3.2.3)

guadabsb15 committed:
added inputs for the regression tests

guadabsb15 committed:
Merge branch 'development' into MHD

guadabsb15 committed:
compile fixes

zingale commented:
Note: our current test suite passes when run on this branch:

http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/

as well as the GPU test suite.

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
remove dup prototype

zingale committed:
separate out MHD prototypes

zingale committed:
Merge branch 'development' into MHD

zingale committed:
add guards

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
fix compilation

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
update to NETWORK_INPUTS

zingale committed:
Merge branch 'development' into MHD

Commit:
Tabs have been converted to spaces by tab_exterminator.sh

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Mhd resete (#820)

update to the new reset_internal_e interface

zingale committed:
Merge branch 'development' into MHD

zingale committed:
fix dupe

zingale committed:
fix interface

zingale committed:
lower small_temp

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'development' into MHD

guadabsb15 committed:
removed ctoprim subroutine from advection_util_nd.F90
to be consistent with development branch, as it somehow appeared again

guadabsb15 committed:
temporarily added the dpdr and dpde variables, to work with  MHD

but will be changing the approach to be consistent with development

guadabsb15 committed:
removed files mhd_normalized_species files

as it is not being called anywhere here and Castro does that on a clean_state

guadabsb15 committed:
removed the local files with enforce_minimum_density

They were not being called and Castro handles that with the clean_state

guadabsb15 committed:
remove hll file

guadabsb15 committed:
remove mhd_trace_ppm

we will focus on plm first, and then if we need ppm we will build up from Castro's one

guadabsb15 committed:
some cleanup

removes gdnv variable, since it was not being used

guadabsb15 committed:
a first pass to ctoprim

one of the thins for the next pass will be making eos_soundspeed smarter

guadabsb15 committed:
more changes in mhd_sound_speed

this seems to fix the crashes seen in the test suite. I will still revise this function later on

zingale committed:
Merge branch 'development' into mhd_merge

guadabsb15 committed:
change in corner_couple

take the directions as parameters so that we don't have to do all of them in the same call of the subroutine

guadabsb15 committed:
fix component

had missed one of the components to be changed to dir

guadabsb15 committed:
another fix

generalized the flux component to be used

zingale committed:
Merge branch 'development' into MHD

guadabsb15 committed:
more ctu refactoring

worked on the half_step subroutine

guadabsb15 committed:
updated wdmerger setup in the mhd_tests

this is just for compilation porpouses, need to later update inputs and
probin

guadabsb15 committed:
fix in half_step

guadabsb15 committed:
refactoring the courner_couple_mag subroutine

guadabsb15 committed:
some more changes to corner_couple

a little simplification and more consistent with corner_couple_mag

guadabsb15 committed:
some more changes in the half_step routine

making it more consistent with the corner_couple routines

guadabsb15 committed:
refactoring half_step_mag

guadabsb15 committed:
get limits in array form instead of components in ctu

guadabsb15 committed:
rearrange some terms and a pass at limits in ctu

There is a lot of room for improving this limits but this is a first pass
Things can be improve if we manage differently the shift of q_p in hlld

guadabsb15 committed:
removed some unused routines

seems like those were for debugging

guadabsb15 committed:
plm clean up

removed the temp arrays, since they are not needed with verified limits

zingale committed:
Merge branch 'development' into MHD

Commit:
Tabs have been converted to spaces by tab_exterminator.sh

guadabsb15 committed:
some fixes

was forgeting to get q(QFS+nspec-1) from U(UFS+nspec-1) in ctoprim
the dimensions of src were NVAR instead of NSRC
fix compilation for wdmerger

zingale committed:
Merge branch 'development' into MHD

zingale committed:
fix compilation

guadabsb15 committed:
fix using B_old for do_old_sources

valgrind was showin unitialized data there when using gravity

guadabsb15 committed:
changes to get hydro_source

getting hydro_source instead of modifying S_new
next chage will include the use of thermo_source

guadabsb15 committed:
finish(ish) transition to construct hydro_source

now using thermo_source for the contribution of -pdivu on UEINT
removed the contribution of SrCQ from plm for now, will add those
contributions soon

Commit:
Tabs have been converted to spaces by tab_exterminator.sh

guadabsb15 committed:
Dust collapse setup

this compiles and the input files run

guadabsb15 committed:
Sod_stellar test

guadabsb15 committed:
RT

a "2D" setup for an x-z RT

guadabsb15 committed:
comments clean up

maxpkatz commented on Source/driver/Derive_F.H:
```diff
@@ -0,0 +1,85 @@
+#endif
+
+#ifdef MHD
+  void ca_dermagcenx
```
These will need to be converted to the new Derive interface.

zingale committed:
Merge branch 'development' into MHD

zingale committed:
Merge branch 'MHD' of github.com:amrex-astro/Castro into MHD

guadabsb15 committed:
inputs to play with for wdmerger

zingale committed:
fix thermo source ifdef

zingale committed:
Merge branch 'MHD' of github.com:amrex-astro/Castro into MHD

guadabsb15 committed:
removed duplicated DustCollapse in mhd_tests

and added necessary inputs and changes to the one in gravity_tests

zingale commented on Source/driver/meth_params.template:
```diff
@@ -17,6 +17,10 @@ module meth_params_module
   ! number of ghost cells for the hyperbolic solver
   integer, parameter     :: NHYP    = 4
+#ifdef MHD
+  integer, parameter     :: MAXADV  = 5
```
I don't think `MAXADV` is used anywhere

guadabsb15 replied:
agree, we don't need it. At least grep suggests that :)

zingale commented on Source/hydro/Castro_hydro.cpp:
```diff
@@ -129,6 +129,12 @@ Castro::cons_to_prim_fourth(const Real time)
+    MultiFab& By = get_new_data(Mag_Type_y);
+    MultiFab& Bz = get_new_data(Mag_Type_z);
+#endif
+
```
are these needed in Castro_hydro.cpp?

guadabsb15 replied:
nope, doesn't seem like we need them there

maxpkatz commented:
I'd prefer if we avoided checking in unclean notebooks into the main repo.

guadabsb15 committed:
removed wdmerger from mhd_tests

added the inputs and probin in the science/wdmerger/tests/wdmerger_mhd

zingale committed:
port the derives to C++ (#896)

guadabsb15 committed:
removing duplicated Sod_stellar

now this test is in hydro_test

guadabsb15 committed:
removed duplicated Sedov

guadabsb15 committed:
have sod inputs in brio&wu

guadabsb15 committed:
removed notebook form DaiWoodward

guadabsb15 committed:
removed dt_cutoff from inputs

guadabsb15 committed:
removed unused variables

guadabsb15 commented:
> For the tests in the MHD directory, can you please remove the tests that are duplicates of existing problem setups like DustCollapse and wdmerger? I'm OK with you adding new inputs files to existing setups, and adding the initmag calls where necessary. That is preferable to maintaining two versions of each setup.

I think it is done
